### ISSUE TITLE
Prevent data loss on reloaded instanced scenes (which use tool scripts)

### ISSUE THREAD

**Issue by Error7Studios** [godotengine/godot-proposals#1012]
**Describe the project you are working on:**
A combat system which makes heavy use of buttons on the screen.

**Describe the problem or limitation you are having in your project:**
I have a custom "BTN" scene with a tool script.
(The BTN node is actually a generic GUI node that has a child Label for text windows and other UI, but can also be clicked.)
I'm instancing a bunch of these BTN nodes to create my game's GUI.
It uses an exported `setget` function to set the text of the child Label in the editor to preview how the scene will look.

The problem is that, because it uses a tool script, whenever I make any change to the script and save it, all the instanced scenes are reloaded and their `setget` setter functions get called again, causing the button text to be reset.
The sizes of the buttons also get reset, so the entire GUI layout gets screwed up.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Automatically re-apply all property settings changes on instanced scenes whenever they are reloaded.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I found a workaround to this problem by doing the following:

1. Set the Main scene (which contains the instanced BTNs) as a tool script.
2. In the Main.gd _ready() function, grab its .tscn file using a Directory search. (_see Notes section_)
3. Parse the text of the Main.tscn file and extract all of the saved properties of each instanced BTN.
4. Find and grab each BTN in the Main scene using get_node() and the node paths in the Main.tscn file.
5. Set each of those BTNs' properties again using the data found in the .tscn file.
6. Call setter functions for each property, if they have one. (_see Notes section_)

**Things that didn't work:**
- Attempts to save the BTNs' data into a dictionary with instance ID keys failed because when the scenes reloaded, each BTN got a new instance ID.
- Saving the data into a non-tool script didn't work since the tool script would run before the non-tool script.
- Saving the data into another tool script didn't work because those scripts also got reset upon restarting the engine.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This would be benefit anyone that uses tool nodes to lay out a scene.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This requires a lot of complex code to do manually.
The fact that nodes can lose their data because they use tool scripts could be a huge source of frustration for people.
Users want something that "just works", and should be able to quickly set up a scene in the editor for rapid prototyping/testing.
They shouldn't have to worry about their setup being reset for inexplicable reasons and losing hours (or more) of work.
This had me pulling my hair out until I implemented my workaround (which isn't perfect.)

**Notes:**
Step 2 (grabbing a scene's .tscn file) was not as simple as I had hoped.
A function like  **Node.get_tscn()** would have been helpful. (This could be a separate proposal)
My method currently relies on the .tscn file having the same name as the root node, and there not being another .tscn file of the same name.

Step 6 currently relies on the setter function being named "set_" + <property_name>, but obviously not everyone names their setters that way.
If there is a built-in way to check if a property has a setter, so that it can be called automatically instead of directly setting the value, that would be useful here.
(Vnen's proposal #844 might solve this problem, but I'm not sure.)

I can provide an example project with this code if it will help, but it might be too specific to my project, and there's probably a much better way to do this than my workaround.

**Comment by Zylann** [comment]
This is a wide problem with `tool` scripts that I encounter every day in plugin development. For some reason, when a tool script reloads because it was modified, ALL its member variables get reset to defaults (unless maybe exported ones, but these are quite rare in my case).

There is no way hot-reloading scripts can work flawlessly. It will break to some extent, but there are obvious ways to limit that.

What I wish would happen is for the script instance's member variables to be preserved temporarily, and re-applied after reload. So none of the references and values they were holding will be lost, which would make logic modification possible without an issue. This goes beyond the `STORAGE` hint of exported variables (it should not rely on that) because `onready` variables and other members referring to nodes and other states also get reset in that process. Also, implementation is language-specific, as member vars may be stored differently (other languages may require similar specific implementation).

Another approach is to reload the scene, but if you do that you loose the point of reloading, and states will remain reset. Invoking `_ready` again was mentionned too, however that only applies to nodes and makes the function multi-purposed, which is harder to use correctly.

**Comment by Error7Studios** [comment]
> What I wish would happen is for the script instance's member variables to be preserved temporarily, and re-applied after reload, without having to intrusively modify the script to take care of that somehow.

It seems like for every script you create, there needs to be a linked backup script created automatically by the engine that's hidden from the user.

Whenever you save a scene or make a change to a script, the backup scripts should get updated as well.

Then, whenever the scene gets reloaded, the engine could check each reloaded script, and compare it to the backup.

Any properties in the original scripts that are different from the ones in the backup since the last save should get set to their value in the backup, and call any relevant setter functions.

I have no idea how much work that would entail, but I don't know how else this could be handled.

What I did was put the .tscn parser code into a function in a singleton, and just called that function in the beginning of Main's `_ready()` func, but that's just a workaround, and only works for my particular use-case (for now).

**Comment by Zylann** [comment]
Yeah that's the idea. It doesnt have to be a "backup script", but may contain the "state" of members so they can be restored or discarded. You basically did that manually, though only on exported properties.

**Comment by KoBeWi** [comment]
I forgot about this, but in 4.0 scripts no longer lose data when saved. Can you check if it resolves the proposal?

**Comment by DissonantVoid** [comment]
what about godot 3.x? is it possible for whatever fixed this be back-ported?
